src/charm.py

```diff
@@ -297,27 +331,39 @@ def _on_ceph_admin_joined(self, event: RelationJoinedEvent) -> None:  # pragma:
             # created resources
             return
 
-        unit_data = event.relation.data[event.unit]
-        expected_relation_keys = (
-            "fsid",
-            "key",
-            "mon_hosts",
-        )
-
-        for relation_key in expected_relation_keys:
-            self._stored.ceph_data[relation_key] = unit_data.get(relation_key)
+        if self.safe_load_ceph_admin_data(event.relation, event.unit):
+            all_resources = self.render_all_resource_definitions()
+            self.create_ceph_resources(all_resources)
+            self._stored.resources_created = True
 
-        missing_data = [key for key, value in self._stored.ceph_data.items() if value is None]
-        if missing_data:
-            logger.warning(
-                "Ceph relation with %s is missing data: %s", event.unit.name, missing_data
-            )
-            self.unit.status = WaitingStatus("Ceph relation is missing data.")
+    def _on_ceph_admin_changed(self, event: RelationChangedEvent) -> None:
+        self.check_required_relations()
+        if not self.unit.is_leader():
+            # Skip resource update on non-leader unit
+            logger.info("Skipping Kubernetes resource update from non-leader unit")
             return
 
-        all_resources = self.render_all_resource_definitions()
-        self.create_ceph_resources(all_resources)
-        self._stored.resources_created = True
+        if self.safe_load_ceph_admin_data(event.relation, event.unit):
+            if self._stored.resources_created:
+                # Update already existing resources that depend on ceph-admin data
+                secret_key = self.ceph_context["kubernetes_key"]
+                fsid = self.ceph_context["fsid"]
+                mon_hosts = self._stored.ceph_data.get("mon_hosts", "").split()
+
+                for resource in self.resources:
+                    if isinstance(resource, Secret):
+                        resource.update_opaque_data("userKey", secret_key)
+                    elif isinstance(resource, StorageClass):
+                        resource.update_cluster_id(fsid)
+                    elif isinstance(resource, ConfigMap):
+                        config_data = [{"clusterID": fsid, "monitors": mon_hosts}]
+                        resource.update_config_json(json.dumps(config_data, indent=4))
```

Comment:
What do you think about adding debug log in `else`?

Reply:
Not sure what would be the benefit. It'd just list every other k8s resource managed by this charm.